File: blocks-world/src/main/scala-2.11/com/btesila/aose/lab7/Predicate.scala
package com.btesila.aose.lab7

trait Predicate { self: Arm =>

  def on(A: Block, B: Block): Boolean = {
    for ( i <- 0 to currentStack.length - 2) {
      if (currentStack(i).equals(A) && currentStack(i + 1).equals(B)) {
        return true
      }
    }
    false
  }

  def clear(A: Block): Boolean =
    currentStack(0).equals(A)

  def onTable(A: Block): Boolean =
    currentStack(currentStack.length - 1).equals(A)

  def hold(A: Block): Boolean =
    self.currentBlock.isDefined && self.currentBlock.get.equals(A)

  def emptyArm(): Boolean =
    self.currentBlock.isEmpty
}

File: blocks-world/src/main/java/com/btesila/aose/lab6/Predicate.java
package com.btesila.aose.lab6;//*H============================================================================
//*M                                EADS RESTRICTED
//*H============================================================================
//*H
//*S  $$Id: Created on 2016.11.07 by ramona.popa@airbus.com $$
//*H
//*C  COPYRIGHT
//*C  This software is copyrighted. It is the property of EADS Deutschland
//*C  GmbH which reserves all right and title to it. It must not be reproduced,
//*C  copied, published or released to third parties nor may the content
//*C  be disclosed to third parties without the prior written consent of
//*C  EADS Deutschland GmbH. Offenders are liable to the payment of damages.
//*C  All rights reserved in the event of the granting of a patent or the
//*C  registration of a utility model or design.
//*C  (c) EADS Deutschland GmbH 2016
//*H
//*H============================================================================

import java.util.Stack;

/**
 * TODO Add class description.
 *
 * @author ramona.popa@airbus.com (original version)
 * @author ramona.popa@airbus.com (responsible)
 * @since 2016.11.07 @ 11:24
 */
public abstract class Predicate {

    public abstract boolean on(Block A, Block B) ;


    public abstract boolean onTable(Block A);

    public abstract boolean  clear(Block A);


    public abstract Boolean hold(Block A);

    public abstract Boolean armEmpty();

}